transports/tcp/src/lib.rs

```diff
@@ -22,362 +22,597 @@
+                                    if addrs.insert(ip) {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        log::debug!("New listen address: {}", ma);
+                                        me.port_reuse.register(ip, me.listen_addr.port());
+                                        return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(ma))));
+                                    }
+                                }
+                            }
+                            Ok(IfEvent::Down(inet)) => {
+                                let ip = inet.addr();
+                                if me.listen_addr.is_ipv4() == ip.is_ipv4() {
+                                    if addrs.remove(&ip) {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        log::debug!("Expired listen address: {}", ma);
+                                        me.port_reuse.unregister(ip, me.listen_addr.port());
+                                        return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(ma))));
+                                    }
+                                }
+                            }
+                            Err(e) => {
+                                me.disable_port_reuse();
+                                return Poll::Ready(Some(Err(e)))
+                            }
+                        }
+                    },
+                },
+                // If the listener is bound to a single interface, make sure the
+                // address is registered for port reuse and reported once.
+                InAddr::One { addr, out } => if let Some(multiaddr) = out.take() {
+                    me.port_reuse.register(*addr, me.listen_addr.port());
+                    return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(multiaddr))))
+                }
+            }
 
-    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        AsyncWrite::poll_flush(Pin::new(&mut self.inner), cx)
-    }
+            if let Some(mut pause) = me.pause.take() {
+                match Pin::new(&mut pause).poll(cx) {
+                    Poll::Ready(_) => {}
+                    Poll::Pending => {
+                        me.pause = Some(pause);
+                        return Poll::Pending;
+                    }
+                }
+            }
 
-    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        AsyncWrite::poll_close(Pin::new(&mut self.inner), cx)
-    }
-}
+            // Check if the listener is ready to accept a new connection.
+            match me.listener.poll_readable(cx) {
+                Poll::Pending => return Poll::Pending,
+                Poll::Ready(Ok(())) => {}
+                Poll::Ready(Err(err)) => {
+                    me.disable_port_reuse();
```

Comment:
in this case it's fine to disable, but I'd prefer matching the error to make sure it is indeed caused by port reuse

Reply:
Similarly to the other case, if `poll_readable()` errors for whatever reason, it is currently considered a fatal error for the `TcpListenStream` and hence de-registering the listen addresses from port reuse. If you think there are some error cases of `poll_readable()` that shouldn't be fatal, we could emit a `ListenerEvent::Error` instead and not call `disable_port_reuse()`. I'd just prefer to be conservative and only treat certain errors as non-fatal if we're certain that the `TcpListenStream` can continue to make progress after such errors.